util/src/main/java/org/cryptomator/util/crypto/HubDeviceCryptor.java

```diff
@@ -0,0 +1,219 @@
+		Preconditions.checkArgument(jwe.getState() == JWEObject.State.DECRYPTED);
+		var fields = jwe.getPayload().toJSONObject();
+		if (fields == null) {
+			Timber.tag("HubDeviceCryptor").e("Expected JWE payload to be JSON: " + jwe.getPayload());
+			throw new MasterkeyLoadingFailedException("Expected JWE payload to be JSON");
+		}
+		var keyBytes = new byte[0];
+		try {
+			if (fields.get(keyField) instanceof String key) {
+				keyBytes = Base64.getDecoder().decode(key);
+				return rawKeyFactory.apply(keyBytes);
+			} else {
+				throw new IllegalArgumentException("JWE payload doesn't contain field " + keyField);
+			}
+		} catch (IllegalArgumentException | KeyDecodeFailedException e) {
+			Timber.tag("HubDeviceCryptor").e("Unexpected JWE payload: " + jwe.getPayload());
+			throw new MasterkeyLoadingFailedException("Unexpected JWE payload", e);
+		} finally {
+			Arrays.fill(keyBytes, (byte) 0x00);
+		}
+	}
+
+	private static JWEObject encryptKey(Key key, ECPublicKey userKey) {
+		try {
+			var encodedVaultKey = Base64.getEncoder().encodeToString(key.getEncoded());
+			var keyGen = new ECKeyGenerator(Curve.P_384);
+			var ephemeralKeyPair = keyGen.generate();
+			var header = new JWEHeader.Builder(JWEAlgorithm.ECDH_ES, EncryptionMethod.A256GCM).ephemeralPublicKey(ephemeralKeyPair.toPublicJWK()).build();
+			var payload = new Payload(Map.of(JWE_PAYLOAD_KEY_FIELD, encodedVaultKey));
+			var jwe = new JWEObject(header, payload);
+			jwe.encrypt(new ECDHEncrypter(userKey));
+			return jwe;
+		} catch (JOSEException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static JWEObject encryptUserKey(ECPrivateKey userKey, ECPublicKey deviceKey) {
+		return encryptKey(userKey, deviceKey);
+	}
+
+	public static Masterkey decryptVaultKey(JWEObject jwe, ECPrivateKey privateKey) throws InvalidJweKeyException {
+		try {
+			jwe.decrypt(new ECDHDecrypter(privateKey));
+			return readKey(jwe, JWE_PAYLOAD_KEY_FIELD, Masterkey::new);
+		} catch (JOSEException e) {
+			throw new InvalidJweKeyException(e);
+		}
+	}
+
+	public JWEObject encryptUserKey(JWEObject userKey, String setupCode) {
+		var userPrivateKey = decryptUserKey(userKey, setupCode);
+		var devicePublicKey = getDevicePublicKey();
+		return encryptUserKey(userPrivateKey, devicePublicKey);
+	}
+
+	public Masterkey decryptVaultKey(JWEObject vaultKeyJwe, JWEObject userKeyJwe) {
+		try {
+			var privateKey = (PrivateKey) keyStore.getKey(DEFAULT_KEY_ALIAS, null);
+			var userKey = decryptUserKey(userKeyJwe, privateKey);
+			return decryptVaultKey(vaultKeyJwe, userKey);
+		} catch (KeyStoreException | NoSuchAlgorithmException | UnrecoverableKeyException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public ECPublicKey getDevicePublicKey() {
+		try {
+			var certificate = keyStore.getCertificate(DEFAULT_KEY_ALIAS);
+			return (ECPublicKey) certificate.getPublicKey();
+		} catch (KeyStoreException e) {
+			throw new RuntimeException(e);
+		}
```

Comment:
Same here:
> I'd suggest choosing a different exception for wrapping `KeyStoreException`.

Reply:
This is the same handling as our other keystore access classes. Lets create an issue and refactor all occurrences later.